### PR TITLE
ISPN-11611 Add internal metadata to persistence

### PR DESCRIPTION
**Pull request by pruivo**
https://issues.redhat.com/browse/ISPN-11611

Depends on: https://github.com/infinispan/infinispan/pull/8113

### CODE REVIEW

**Comment by ryanemerson** [comment]
@wburns Any thoughts on how best to maintain compatibility with the file stores?

File: core/src/main/java/org/infinispan/persistence/file/SingleFileStore.java [hunk 1]
```diff
@@ -196,11 +197,12 @@ private void rebuildIndex() throws Exception {
          int keyLen = buf.getInt();
          int dataLen = buf.getInt();
          int metadataLen = buf.getInt();
+         int internalMetadataLen = buf.getInt();
```
[the remark this answers is not in the document]
Comment (pruivo) [reply core/src/main/java/org/infinispan/persistence/file/SingleFileStore.java:200]: I agree the MAGIC should be changed... but does it need to be backward compatible? 
isn't that the StoreMigrator's job?

[the remark this answers is not in the document]
Comment (ryanemerson) [reply core/src/main/java/org/infinispan/persistence/file/SingleFileStore.java:200]: The `StoreMigrator` is a get out of jail, not quite for free, card. The long term goal is to get so that we can do rolling upgrades seamlessly in kubernetes, so we need no manual steps.

File: core/src/main/java/org/infinispan/eviction/impl/PassivationManagerImpl.java [hunk 2]
```diff
@@ -72,10 +71,9 @@ private boolean isL1Key(Object key) {
       return distributionManager != null && !distributionManager.getCacheTopology().isWriteOwner(key);
    }
 
-   private CompletionStage<Boolean> doPassivate(Object key, CacheEntry<?, ?> entry) {
+   private CompletionStage<Boolean> doPassivate(Object key, InternalCacheEntry<?, ?> entry) {
       if (trace) log.tracef("Passivating entry %s", toStr(key));
-         MarshallableEntry<?, ?> marshalledEntry = marshalledEntryFactory.create(key, entry.getValue(), entry.getMetadata(),
-               entry.getCreated(), entry.getLastUsed());
+      MarshallableEntry marshalledEntry = marshalledEntryFactory.create((InternalCacheEntry) entry);
```
[the remark this answers is not in the document]
Comment (pruivo) [reply core/src/main/java/org/infinispan/eviction/impl/PassivationManagerImpl.java:76]: the cast is necessary for some weird reason that I fail to understand. 
If you don't cast, java will invoke `org.infinispan.persistence.spi.MarshallableEntryFactory#create(java.lang.Object)`

File: persistence/rocksdb/src/main/java/org/infinispan/persistence/rocksdb/RocksDBStore.java [hunk 3]
```diff
@@ -309,25 +309,21 @@ public void deleteBatch(Iterable<Object> keys) {
...
+                expiredDb.put(expiryBytes, marshall(existing));
             } else {
-                expiredDb.put(expiryBytes, entry.keyBytes);
+                ExpiryBucket bucket = new ExpiryBucket(existingBytes, entry.keyBytes);
+                expiredDb.put(expiryBytes, marshall(bucket));
             }
-        } catch (IOException | InterruptedException | RocksDBException | ClassNotFoundException e) {
-            throw e;
+        } else {
+            expiredDb.put(expiryBytes, entry.keyBytes);
```
[the remark this answers is not in the document]
Comment (pruivo) [reply persistence/rocksdb/src/main/java/org/infinispan/persistence/rocksdb/RocksDBStore.java:326]: no, but catching an exception just to throw it, is a waste of bytecodes :P

File: core/src/main/java/org/infinispan/persistence/file/SingleFileStore.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (pruivo) [reply core/src/main/java/org/infinispan/persistence/file/SingleFileStore.java:200]: oh... that was unexpected... 
SingleFileStore can be fixed but the remaining stores (RocksDB, JDBC, SIFS, JPA) are probably broken. there is no "magic" or "version" field there :/

[the remark this answers is not in the document]
Comment (danberindei) [reply core/src/main/java/org/infinispan/persistence/file/SingleFileStore.java:200]: @pruivo we can relatively easy add a new column family for metadata in RocksDB, and a metadata table in JDBC/JPA. For SIFS we may need a breaking change.

[the remark this answers is not in the document]
Comment (pruivo) [reply core/src/main/java/org/infinispan/persistence/file/SingleFileStore.java:200]: @danberindei not sure if it is going to work. JDBC and RockDB store the `MarshalledValue` as `byte[]`. I guess protostream will fail to unmarshall it because it expects to have another field... 

About JPA, my knowledge is -infinity. I just copy what others did
```
@Lob
@Column(length = 65535)
private byte[] internalMetadata;
```

[the remark this answers is not in the document]
Comment (danberindei) [reply core/src/main/java/org/infinispan/persistence/file/SingleFileStore.java:200]: My bad, I was only thinking about the missing magic/version fields, so was using "metadata" as "information about the store itself", not as in "information about the entries".

I'm not very familiar with JPA either, but for JDBC/RocksDB it should be fairly easy to check when opening a store if the new table/column family exists or not.

However, this doesn't completely solve our problem, we also have to migrate the existing entries, or we wouldn't be able to change the version of the store, and we wouldn't be able to store the internal metadata.

File: persistence/soft-index/src/main/java/org/infinispan/persistence/sifs/EntryHeader.java [hunk 4]
```diff
@@ -23,6 +24,7 @@ public EntryHeader(ByteBuffer buffer) {
       this.keyLength = buffer.getShort();
       this.metadataLength = buffer.getShort();
       this.valueLength = buffer.getInt();
+      this.internalMetadataLength = buffer.getInt();
```
[the remark this answers is not in the document]
Comment (danberindei) [reply persistence/soft-index/src/main/java/org/infinispan/persistence/sifs/EntryHeader.java:27]: I would prefer adding the version information as a separate file, so we can easily check if the version information exists and default to the 10.1 layout (or migrate in-place to the 11+ layout).

File: core/src/main/java/org/infinispan/persistence/file/SingleFileStore.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (ryanemerson) [reply core/src/main/java/org/infinispan/persistence/file/SingleFileStore.java:200]: @pruivo @danberindei  For store implementations that use `MarshalledValueImpl` to store the value, i.e. Jdbc & Rocksdb, we should be ok because you have added the internal metadata there. The way Protobuf/ProtoStream works, if a field is missing from the old bytes, then it just isn't read and so in `MarshalledValueImpl` we can just set the internal metadata to null. So there shouldn't be any additional logic in the store itself AFAICT.

[the remark this answers is not in the document]
Comment (ryanemerson) [reply core/src/main/java/org/infinispan/persistence/file/SingleFileStore.java:200]: > About JPA, my knowledge is -infinity. I just copy what others did

That's what we all do :laughing:

File: persistence/soft-index/src/main/java/org/infinispan/persistence/sifs/EntryHeader.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (ryanemerson) [reply persistence/soft-index/src/main/java/org/infinispan/persistence/sifs/EntryHeader.java:27]: A separate file makes sense.

**Comment by pruivo** [comment]
@ryanemerson @danberindei a small update... I did some tests with data written by 10.1.5.Final and read by this patch:
* `JPA Store`: it works fine. some layer handles it for us
* `RocksDB` and `JDBC Store`: they were throwing NPEs because Protostream sets the new field to `null`. It is fixed now
* `Single File Store`: the `magic` number is written at the beginning of the file so I'm migrating the data when the store starts. The file is replaced.
* `Soft Index File Store`: the format used is based on the file name. The previous format is just a number. I added a prefix, so the new format is `ispn.[0-9]+`. I'm migrating the data when the store starts and I enabled the `magic` number (per entry) so we don't have to do any more migrations in the future.

I've pushed the fixed. Check the new commits :)

File: commons/all/src/main/java/org/infinispan/commons/io/EmptyByteBuffer.java [hunk 5]
```diff
@@ -0,0 +1,36 @@
+package org.infinispan.commons.io;
+
+import org.infinispan.commons.util.Util;
+
+/**
+ * //TODO document this!
```
Comment (ryanemerson) [inline commons/all/src/main/java/org/infinispan/commons/io/EmptyByteBuffer.java:6]: Don't foget :slightly_smiling_face:

File: persistence/soft-index/src/main/java/org/infinispan/persistence/sifs/EntryHeader.java [hunk 6]
```diff
@@ -7,26 +7,33 @@
  */
 public class EntryHeader {
    static final int MAGIC = 0xBE11A61C;
-   static final boolean useMagic = false;
-   static final int HEADER_SIZE = 28 + (useMagic ? 4 : 0);
+   //no magic neither no internal metadata
+   static final int OLD_HEADER_SIZE = 24;
+   static final int HEADER_SIZE = 32;
```
Comment (ryanemerson) [inline persistence/soft-index/src/main/java/org/infinispan/persistence/sifs/EntryHeader.java:12]: `HEADER_SIZE_10_1` and `HEADER_SIZE_11_0` would be more informative.

File: persistence/soft-index/src/main/java/org/infinispan/persistence/sifs/EntryHeader.java [hunk 7]
```diff
@@ -53,12 +60,26 @@ public long expiryTime() {
...
       return String.format("[keyLength=%d, valueLength=%d, metadataLength=%d, internalMetadataLength=%d,seqId=%d, expiration=%d]", keyLength, valueLength, metadataLength, internalMetadataLength, seqId, expiration);
    }
 
    public int totalLength() {
-      return keyLength + metadataLength + internalMetadataLength + valueLength + HEADER_SIZE;
+      return keyLength + metadataLength + internalMetadataLength + valueLength + headerLength;
+   }
+
+   public static void writeHeader(ByteBuffer buf, short keyLength, short metadataLength, int valueLength, int internalMetadataLength, long seqId, long expiration) {
+      buf.putInt(EntryHeader.MAGIC);
```
Comment (ryanemerson) [inline persistence/soft-index/src/main/java/org/infinispan/persistence/sifs/EntryHeader.java:77]: I would prefer to avoid writing MAGIC per header if possilbe. However, if we can't avoid that we should at least store it as a single byte. `useMagic` was only ever used for testing, so we don't have to worry about the compatibility of this field as it was never used in production.

File: tools/src/main/java/org/infinispan/tools/store/migrator/file/SoftIndexFileStoreIterator.java [hunk 8]
```diff
@@ -85,7 +85,7 @@ public MarshallableEntry next() {
                      byte[] serializedValue = EntryRecord.readValue(handle, header, offset);
 
                      offset += header.totalLength();
-                     if (EntryRecord.readEntryHeader(handle, offset) == null) {
+                     if (EntryRecord.readOldEntryHeader(handle, offset) == null) {
```
Comment (ryanemerson) [inline tools/src/main/java/org/infinispan/tools/store/migrator/file/SoftIndexFileStoreIterator.java:88]: What if the user is migrating a store from the existing version of Infinispan, i.e. 11.x?

File: persistence/soft-index/src/main/java/org/infinispan/persistence/sifs/EntryHeader.java (hunk 6 above)
Comment (pruivo) [reply persistence/soft-index/src/main/java/org/infinispan/persistence/sifs/EntryHeader.java:12]: ack

File: commons/all/src/main/java/org/infinispan/commons/io/EmptyByteBuffer.java (hunk 5 above)
Comment (pruivo) [reply commons/all/src/main/java/org/infinispan/commons/io/EmptyByteBuffer.java:6]: ops...

File: persistence/soft-index/src/main/java/org/infinispan/persistence/sifs/EntryHeader.java (hunk 7 above)
Comment (pruivo) [reply persistence/soft-index/src/main/java/org/infinispan/persistence/sifs/EntryHeader.java:77]: I'll change to byte. 
IMO, it is better to keep a `magic` number per entry. If I understand correctly, SIFS only appends the keys changes to a new file; So, if the format changes in the future, we don't need any migration during startup.

File: tools/src/main/java/org/infinispan/tools/store/migrator/file/SoftIndexFileStoreIterator.java (hunk 8 above)
Comment (pruivo) [reply tools/src/main/java/org/infinispan/tools/store/migrator/file/SoftIndexFileStoreIterator.java:88]: no sure If I understand what you mean... migrating for 11.0 to 11.x? 

The filenames changed with this PR and the FileProvider only searches for old file names, i.e. the file names previous to this PR.
After this PR, the users don't need to migrate anything. We have the magic number per entry so we are able to read anything.

**Comment by pruivo** [comment]
@ryanemerson should I store the length of `internalMetadata` as `short`? 
The `key` and `metadata` length are limited to a `short`.

**Comment by ryanemerson** [comment]
> @ryanemerson should I store the length of `internalMetadata` as `short`?
> The `key` and `metadata` length are limited to a `short`.

I'm all for saving space ... but I'm not sure if it is safe to store these lengths as shorts. Throughout all of our code we assume `int`, so this might be an issue that so far has gone unnoticed. @danberindei @rvansa any thoughts?

**Comment by pruivo** [comment]
The keys are limited to 32KB: https://github.com/infinispan/infinispan/blob/master/persistence/soft-index/src/main/java/org/infinispan/persistence/sifs/SoftIndexFileStore.java#L85

The internal metadata should be small enough to fit in a short.

File: tools/src/main/java/org/infinispan/tools/store/migrator/file/SoftIndexFileStoreIterator.java (hunk 8 above)
Comment (ryanemerson) [reply tools/src/main/java/org/infinispan/tools/store/migrator/file/SoftIndexFileStoreIterator.java:88]: > no sure If I understand what you mean... migrating for 11.0 to 11.x?

Yes. Currently it's possible to migrate from one store implementation to another, e.g. SIFS -> RocksDB, within the latest Infinispan version.

You can determine the version of the source store via `StoreProperties#getMajorVersion` and call `readOldEntryHeader` if `< 11`

Comment (pruivo) [reply tools/src/main/java/org/infinispan/tools/store/migrator/file/SoftIndexFileStoreIterator.java:88]: ok! 
I wasn't aware of that. I need to update SFS too since it is only parsing the old format.

**Comment by pruivo** [comment]
updated!

File: persistence/soft-index/src/main/java/org/infinispan/persistence/sifs/EntryRecord.java [hunk 9]
```diff
@@ -117,17 +150,31 @@ private static int read(FileProvider.Handle handle, ByteBuffer buffer, long posi
...
+      return array == null ? 0 : (short) array.length;
+   }
+
+   private static int bufferLength(org.infinispan.commons.io.ByteBuffer byteBuffer) {
+      return byteBuffer == null ? 0 : byteBuffer.getLength();
+   }
+
+   private static short bufferLengthAsShort(org.infinispan.commons.io.ByteBuffer byteBuffer) {
+      return byteBuffer == null ? 0 : (short) byteBuffer.getLength();
+   }
```
Comment (ryanemerson) [inline persistence/soft-index/src/main/java/org/infinispan/persistence/sifs/EntryRecord.java:167]: I don't think these static methods simplify things. `arrayLength` and `bufferLenth` are only called once as well. I would prefer just to keep the checks inline.

File: persistence/soft-index/src/main/java/org/infinispan/persistence/sifs/SoftIndexFileStore.java [hunk 10]
```diff
@@ -163,40 +164,93 @@ public void start() {
...
-      } else if (configuration.purgeOnStartup()) {
-         log.debug("Not building the index - purge will be executed");
+      boolean migrateData = false;
+
+      if (!configuration.purgeOnStartup()) {
+         // we don't destroy the data on startup
+         // get the old files
+         FileProvider oldFileProvider = new FileProvider(getDataLocation(), configuration.openFilesLimit(), "");
+         if (oldFileProvider.hasFiles()) {
+            log.debug("Migrate old format to new format.");
```
Comment (ryanemerson) [inline persistence/soft-index/src/main/java/org/infinispan/persistence/sifs/SoftIndexFileStore.java:174]: I think this log and a message indicating that the migration has sucessfully completed should be `info` the same as in SFS.

File: persistence/soft-index/src/main/java/org/infinispan/persistence/sifs/SoftIndexFileStore.java [hunk 11]
```diff
@@ -163,40 +164,93 @@ public void start() {
...
+                  if (header == null) {
+                     //end of file. go to next one
+                     break;
                   }
-                  Object key = marshaller.objectFromByteBuffer(serializedKey);
-                  if (trace) {
-                     log.tracef("Loaded %d:%d (seqId %d, expiration %d)", file, offset, seqId, expiration);
+                  MarshallableEntry<Object, Object> entry = readEntry(handle, header, offset, null);
+                  // entry is null if expired or removed (tombstone)
+                  // in both case, we can ignore it.
```
Comment (ryanemerson) [inline persistence/soft-index/src/main/java/org/infinispan/persistence/sifs/SoftIndexFileStore.java:211]: Nitpick. Single line comment

File: persistence/soft-index/src/main/java/org/infinispan/persistence/sifs/SoftIndexFileStore.java (hunk 10 above)
Comment (pruivo) [reply persistence/soft-index/src/main/java/org/infinispan/persistence/sifs/SoftIndexFileStore.java:174]: should the log message be in the class logger or under `PERSISTENCE` logger?

Comment (ryanemerson) [reply persistence/soft-index/src/main/java/org/infinispan/persistence/sifs/SoftIndexFileStore.java:174]: Good question. I think PERSISTENCE. The same applies for SFS

**Comment by pruivo** [comment]
@ryanemerson thanks! I've updated it based on your comments. 
I'll handle the test after lunch.

**Comment by ryanemerson** [comment]
@pruivo Requires a rebase after https://github.com/infinispan/infinispan/pull/8217 was merged.

**Comment by pruivo** [comment]
@ryanemerson Only the JPA store is missing (i.e. the migration test).
You can check the other tests if you want.

**Comment by pruivo** [comment]
@ryanemerson done! I'll squash the commits when you give me the green light. thanks!

File: commons-test/src/main/java/org/infinispan/commons/test/CommonsTestingUtil.java [hunk 12]
```diff
@@ -47,6 +47,17 @@ public static String tmpDirectory() {
...
    }
 
+   /**
+    * Creates a path to a unique (per test) temporary directory. The directory is created in the platform's temp
+    * directory (set by {@code java.io.tmpdir}).
+    *
+    * @param test test that requires this directory.
+    * @return an absolute path
+    */
+   public static String javaTmpDirectory(Class<?> test) {
```
Comment (ryanemerson) [inline commons-test/src/main/java/org/infinispan/commons/test/CommonsTestingUtil.java:57]: Why is this necessary? Can't we just use `tmpDirectory(Class<?> test)`?

File: commons/all/src/main/java/org/infinispan/commons/io/ByteBufferFactoryImpl.java [hunk 13]
```diff
@@ -8,6 +8,6 @@
 
    @Override
    public ByteBuffer newByteBuffer(byte[] b, int offset, int length) {
-      return new ByteBufferImpl(b, offset, length);
+      return b == null || b.length == 0 || length == 0 ? ByteBufferImpl.EMPTY_INSTANCE : new ByteBufferImpl(b, offset, length);
```
Comment (ryanemerson) [inline commons/all/src/main/java/org/infinispan/commons/io/ByteBufferFactoryImpl.java:11]: Can be:

```java
return ByteBufferImpl.create(b, offset, length);`
```

File: core/src/test/java/org/infinispan/persistence/file/SingleFileStoreCompatibilityTest.java [hunk 14]
```diff
@@ -0,0 +1,89 @@
...
+ * @author Pedro Ruivo
+ * @since 11.0
+ */
+@Test(groups = "functional", testName = "persistence.file.SingleFileStoreCompatibilityTest")
+public class SingleFileStoreCompatibilityTest extends PersistenceCompatibilityTest<String> {
+
+   private static final String DATA_10_1 = "10_1_x_sfs_data/sfs-store-cache.dat";
+   private String tmpDirectory;
+
+
```
Comment (ryanemerson) [inline core/src/test/java/org/infinispan/persistence/file/SingleFileStoreCompatibilityTest.java:28]: Nitpick. Unnecessary empty line

File: core/src/test/java/org/infinispan/persistence/file/SingleFileStoreCompatibilityTest.java [hunk 15]
```diff
@@ -0,0 +1,89 @@
...
+ * Tests if {@link SingleFileStore} can migrate data from Infinispan 10.1.x.
+ *
+ * @author Pedro Ruivo
+ * @since 11.0
+ */
+@Test(groups = "functional", testName = "persistence.file.SingleFileStoreCompatibilityTest")
+public class SingleFileStoreCompatibilityTest extends PersistenceCompatibilityTest<String> {
+
+   private static final String DATA_10_1 = "10_1_x_sfs_data/sfs-store-cache.dat";
+   private String tmpDirectory;
```
Comment (ryanemerson) [inline core/src/test/java/org/infinispan/persistence/file/SingleFileStoreCompatibilityTest.java:26]: We can pass this to the constructor of `PersistenceCompatibilityTest` so that we don't have to override `setup` and `teardown` in all of the tests.

File: core/src/test/java/org/infinispan/persistence/PersistenceCompatibilityTest.java [hunk 16]
```diff
@@ -0,0 +1,128 @@
...
+         assertEquals("Wrong value read for key " + key, value(i), valueWrapper.unwrap(cache.get(key)));
+      }
+
+   }
+
+   protected abstract void beforeStartCache() throws Exception;
+
+   @Override
+   protected EmbeddedCacheManager createCacheManager() throws Exception {
+      GlobalConfigurationBuilder builder = new GlobalConfigurationBuilder().nonClusteredDefault();
```
Comment (ryanemerson) [inline core/src/test/java/org/infinispan/persistence/PersistenceCompatibilityTest.java:90]: We can set the Global persistent location here:

`builder.globalState().persistentLocation(CommonsTestingUtil.tmpDirectory())`

That way not all of the tests have to override `amendGlobalConfigurationBuilder` and it should not negatively effect the JPA and JDBC tests that don't need it.

File: persistence/soft-index/src/main/java/org/infinispan/persistence/sifs/SoftIndexFileStore.java [hunk 17]
```diff
@@ -150,7 +151,7 @@ public void start() {
       temporaryTable = new TemporaryTable(configuration.indexQueueLength() * configuration.indexSegments());
       storeQueue = new SyncProcessingQueue<>();
       indexQueue = new IndexQueue(configuration.indexSegments(), configuration.indexQueueLength());
-      fileProvider = new FileProvider(getDataLocation(), configuration.openFilesLimit());
+      fileProvider = new FileProvider(getDataLocation(), configuration.openFilesLimit(), "ispn.");
```
Comment (ryanemerson) [inline persistence/soft-index/src/main/java/org/infinispan/persistence/sifs/SoftIndexFileStore.java:154]: It would be good to have the prefix values as constants, e.g. 10_1_PREFIX = "", 11_0_PREFIX="ispn."

File: persistence/soft-index/src/main/java/org/infinispan/persistence/sifs/SoftIndexFileStore.java [hunk 18]
```diff
@@ -163,40 +164,95 @@ public void start() {
...
-      } else if (configuration.purgeOnStartup()) {
-         log.debug("Not building the index - purge will be executed");
+      boolean migrateData = false;
+
+      if (!configuration.purgeOnStartup()) {
+         // we don't destroy the data on startup
+         // get the old files
+         FileProvider oldFileProvider = new FileProvider(getDataLocation(), configuration.openFilesLimit(), "");
+         if (oldFileProvider.hasFiles()) {
+            org.infinispan.util.logging.Log.PERSISTENCE.info("Migrating data new format...");
```
Comment (ryanemerson) [inline persistence/soft-index/src/main/java/org/infinispan/persistence/sifs/SoftIndexFileStore.java:174]: We should use the statements you introduced in `Log.java`

Comment (ryanemerson) [reply persistence/soft-index/src/main/java/org/infinispan/persistence/sifs/SoftIndexFileStore.java:174]: The code will be a lot cleaner if we statically import `PERSISTENCE` log.

File: commons-test/src/main/java/org/infinispan/commons/test/CommonsTestingUtil.java (hunk 12 above)
Comment (pruivo) [reply commons-test/src/main/java/org/infinispan/commons/test/CommonsTestingUtil.java:57]: This is for the JPA store. 
The path has to be set in `persistence.xml` but the `tmpDirectory()` method uses `infinispan.test.tmpdir`. If `infinispan.test.tmpdir` is not set, it falls back to `java.io.tmpdir`. 
If I use `infinispan.test.tmpdir` and it was not set, the DB will point to the incorrect location.

File: commons/all/src/main/java/org/infinispan/commons/io/ByteBufferFactoryImpl.java (hunk 13 above)
Comment (pruivo) [reply commons/all/src/main/java/org/infinispan/commons/io/ByteBufferFactoryImpl.java:11]: there is no `ByteBufferImpl.create(b, offset, length);` method. Are you suggesting to create one?

File: core/src/test/java/org/infinispan/persistence/file/SingleFileStoreCompatibilityTest.java (hunk 14 above)
Comment (pruivo) [reply core/src/test/java/org/infinispan/persistence/file/SingleFileStoreCompatibilityTest.java:28]: :+1:

File: core/src/test/java/org/infinispan/persistence/file/SingleFileStoreCompatibilityTest.java (hunk 15 above)
Comment (pruivo) [reply core/src/test/java/org/infinispan/persistence/file/SingleFileStoreCompatibilityTest.java:26]: :+1:

File: core/src/test/java/org/infinispan/persistence/PersistenceCompatibilityTest.java (hunk 16 above)
Comment (pruivo) [reply core/src/test/java/org/infinispan/persistence/PersistenceCompatibilityTest.java:90]: :+1:

File: persistence/soft-index/src/main/java/org/infinispan/persistence/sifs/SoftIndexFileStore.java (hunk 17 above)
Comment (pruivo) [reply persistence/soft-index/src/main/java/org/infinispan/persistence/sifs/SoftIndexFileStore.java:154]: :+1:

File: persistence/soft-index/src/main/java/org/infinispan/persistence/sifs/SoftIndexFileStore.java (hunk 18 above)
Comment (pruivo) [reply persistence/soft-index/src/main/java/org/infinispan/persistence/sifs/SoftIndexFileStore.java:174]: :+1:

File: commons/all/src/main/java/org/infinispan/commons/io/ByteBufferFactoryImpl.java (hunk 13 above)
Comment (ryanemerson) [reply commons/all/src/main/java/org/infinispan/commons/io/ByteBufferFactoryImpl.java:11]: Sorry, I was thinking there was. But yeah lets add `ByteBufferImpl.create(b, offset, length)`, that way all of the logic around `EMPTY_INSTANCE` is in one place and `ByteBufferImpl.create(b)` impl can just call the overloaded method.

Comment (pruivo) [reply commons/all/src/main/java/org/infinispan/commons/io/ByteBufferFactoryImpl.java]: done! 
I changed `PersistenceMarshallerImpl` a little since it was allocating and copying the array twice.

Comment (ryanemerson) [reply commons/all/src/main/java/org/infinispan/commons/io/ByteBufferFactoryImpl.java]: Good spot, thanks

**Comment by ryanemerson** [comment]
Thanks @pruivo
